src/com/google/javascript/jscomp/PeepholeReplaceKnownMethods.java

```diff
@@ -956,4 +959,148 @@ private Node tryFoldStringSplit(Node n, Node stringNode, Node arg1) {
     reportChangeToEnclosingScope(parent);
     return arrayOfStrings;
   }
+
+  private Node tryToFoldArrayConcat(Node n) {
+    checkArgument(n.isCall(), n);
+
+    if (!isASTNormalized() || !useTypes) {
+      return n;
+    }
+    ConcatFunctionCall concatFunctionCall = createConcatFunctionCallForNode(n);
+    if (concatFunctionCall == null) {
+      return n;
+    }
+    concatFunctionCall = tryToremoveArrayLiteralFromFrontOfConcat(concatFunctionCall);
+    checkNotNull(concatFunctionCall);
+    if (isSafeToFoldConcatChaining(concatFunctionCall)) {
```

Comment:
The combined check-then-do-it logic looks great for `tryToRemoveArrayLiteralFromFrontOfConcat()`

Could you do the same here?